htlcswitch/link.go

```diff
@@ -538,15 +627,186 @@ func (l *channelLink) WaitForShutdown() {
 // actively accept requests to forward HTLC's. We're able to forward HTLC's if
-// we know the remote party's next revocation point. Otherwise, we can't
-// initiate new channel state. We also require that the short channel ID not be
-// the all-zero source ID, meaning that the channel has had its ID finalized.
+// we are eligible to update AND the channel isn't currently flushing the
+// outgoing half of the channel.
 func (l *channelLink) EligibleToForward() bool {
+	return l.EligibleToUpdate() &&
+		!l.IsFlushing(Outgoing)
+}
+
+// EligibleToUpdate returns a bool indicating if the channel is able to update
+// channel state. We're able to update channel state if we know the remote
+// party's next revocation point. Otherwise, we can't initiate new channel
+// state. We also require that the short channel ID not be the all-zero source
+// ID, meaning that the channel has had its ID finalized.
+func (l *channelLink) EligibleToUpdate() bool {
 	return l.channel.RemoteNextRevocation() != nil &&
 		l.ShortChanID() != hop.Source &&
 		l.isReestablished()
 }
 
+// EnableAdds sets the ChannelUpdateHandler state to allow UpdateAddHtlc's in
+// the specified direction. It returns an error if the state already allowed
+// those adds.
+func (l *channelLink) EnableAdds(linkDirection LinkDirection) error {
+	if linkDirection == Outgoing {
+		if !l.isOutgoingAddBlocked.Swap(false) {
+			return errors.New("outgoing adds already enabled")
+		}
+	}
+
+	if linkDirection == Incoming {
+		if !l.isIncomingAddBlocked.Swap(false) {
+			return errors.New("incoming adds already enabled")
+		}
+	}
+
+	return nil
+}
+
+// DiableAdds sets the ChannelUpdateHandler state to allow UpdateAddHtlc's in
+// the specified direction. It returns an error if the state already disallowed
+// those adds.
+func (l *channelLink) DisableAdds(linkDirection LinkDirection) error {
+	if linkDirection == Outgoing {
+		if l.isOutgoingAddBlocked.Swap(true) {
+			return errors.New("outgoing adds already disabled")
+		}
+	}
+
+	if linkDirection == Incoming {
+		if l.isIncomingAddBlocked.Swap(true) {
+			return errors.New("incoming adds already disabled")
+		}
+	}
+
+	return nil
+}
+
+// IsFlushing returns true when UpdateAddHtlc's are disabled in the direction of
+// the argument.
+func (l *channelLink) IsFlushing(linkDirection LinkDirection) bool {
+	if linkDirection == Outgoing {
+		return l.isOutgoingAddBlocked.Load()
+	}
+
+	return l.isIncomingAddBlocked.Load()
+}
+
+// OnFlushedOnce is a method that adds a hook to be called the next time the
+// channel state reaches zero htlcs. This hook will only ever be called once.
+// This function returns a FlushHookID that can be used to preemptively remove
+// this hook. If the channel state is already flushed -- no HTLCs and no
+// commitments owed by either party -- then we will call this hook synchronously
+// and will not return a FlushHookID.
+func (l *channelLink) OnFlushedOnce(hook func()) fn.Option[FlushHookID] {
```

Comment:
It's possible that the `hook` function here is never called:

- goroutine 1: `OnFlushedOnce` called and `IsChannelClean` returns false
- goroutine 2: The link receives the final `commit_sig` / `revoke_and_ack` that makes the channel clean in `handleUpstreamMsg`
- goroutine 2: `flushHooks.invoke()` is called, but there are no hooks yet
- goroutine 1: `flushHooks.alloc` is called, but no goroutine will call `flushHooks.invoke()` anymore since if the peer is following the spec, they won't send any more messages

Reply:
Indeed. I'll move the Lock call to the beginning